### PR TITLE
:sparkles: support wildcard in ManifestConfigs

### PR DESCRIPTION
**Pull request by zhiweiyin318**
## Summary
Support wildcard in the name and namespace of ManifestConfigs.ResourceIdentifier for feedbackRule and updateStrategy.
if the a resource matches more than 1 options, only the first one works.
## Related issue(s)
https://github.com/open-cluster-management-io/ocm/issues/705
Fixes #

### CODE REVIEW

File: pkg/work/spoke/controllers/manifestcontroller/manifestwork_reconciler.go [hunk 1]
```diff
@@ -195,7 +195,7 @@ func (m *manifestworkReconciler) applyOneManifest(
 	requiredOwner := manageOwnerRef(ownedByTheWork, owner)
 
 	// find update strategy option.
-	option := helper.FindManifestConiguration(resMeta, workSpec.ManifestConfigs)
+	option := helper.FindManifestConfiguration(resMeta, workSpec.ManifestConfigs)
```
Comment (qiujian16) [inline pkg/work/spoke/controllers/manifestcontroller/manifestwork_reconciler.go:198]: I would want updateStrategy also support wildcard match. I think we should have helper.FindManifestConfigurations(resMeta, workSpec.ManifestConfigs) returns a list. For updateStrategy, we pick the first or latest item in the list, while for feedback rules, we pick every item.

File: test/integration/work/statusfeedback_test.go [hunk 2]
```diff
@@ -242,8 +242,8 @@ var _ = ginkgo.Describe("ManifestWork Status Feedback", func() {
 					ResourceIdentifier: workapiv1.ResourceIdentifier{
 						Group:     "apps",
 						Resource:  "deployments",
-						Namespace: commOptions.SpokeClusterName,
-						Name:      "deploy1",
+						Namespace: "*",
+						Name:      "*1",
```
Comment (qiujian16) [inline test/integration/work/statusfeedback_test.go:246]: we should have some more tests for wildcard case.

File: pkg/work/spoke/controllers/manifestcontroller/manifestwork_reconciler.go (hunk 1 above)
Comment (zhiweiyin318) [reply pkg/work/spoke/controllers/manifestcontroller/manifestwork_reconciler.go]: returned a list of options .

File: test/integration/work/statusfeedback_test.go (hunk 2 above)
Comment (zhiweiyin318) [reply test/integration/work/statusfeedback_test.go:246]: added

Comment (qiujian16) [reply test/integration/work/statusfeedback_test.go:246]: We should also have another test with multiple matched options.

**Comment by qiujian16** [comment]
/approve

@elgnay or @zhujian7 please also take a look

**Comment by zhiweiyin318** [comment]
added more integration tests in feedBackRule and upgradeStrategy : 
1. a wildcard option matches multi resources 2. a resource is matched by multi options. 
@elgnay @zhujian7 please take a look. cc @qiujian16

File: pkg/work/spoke/controllers/manifestcontroller/manifestwork_reconciler.go (hunk 1 above)
Comment (zhujian7) [reply pkg/work/spoke/controllers/manifestcontroller/manifestwork_reconciler.go]: > while for feedback rules, we pick every item.

@zhiweiyin318 want to confirm for the current implementation, do we only pick up the first one?

Comment (zhiweiyin318) [reply pkg/work/spoke/controllers/manifestcontroller/manifestwork_reconciler.go]: fixed. for feedbackRule, will pick all matched options.

File: pkg/work/spoke/controllers/manifestcontroller/manifestwork_reconciler.go [hunk 3]
```diff
@@ -194,16 +194,25 @@ func (m *manifestworkReconciler) applyOneManifest(
...
+	// strategy is updated by default
 	strategy := workapiv1.UpdateStrategy{Type: workapiv1.UpdateStrategyTypeUpdate}
-	if option != nil && option.UpdateStrategy != nil {
-		strategy = *option.UpdateStrategy
+
+	// only handle the first option if the resource matches more than one option.
+	// need avoid a resource from being matched to more options.
+	option := workapiv1.ManifestConfigOption{}
+	if len(options) > 0 {
+		option = options[0]
```
[the remark this answers is not in the document]
Comment (zhiweiyin318) [reply pkg/work/spoke/controllers/manifestcontroller/manifestwork_reconciler.go:207]: fixed. for feedbackRule, will pick all matched options.

[the remark this answers is not in the document]
Comment (qiujian16) [reply pkg/work/spoke/controllers/manifestcontroller/manifestwork_reconciler.go:207]: I am not sure it works. Each feedback result have a unique name, and it should not be overlapped.

**Review by qiujian16 — COMMENTED** [review]
/approve
/lgtm

File: test/integration/work/statusfeedback_test.go (hunk 2 above)
Comment (zhiweiyin318) [reply test/integration/work/statusfeedback_test.go:246]: done

File: pkg/work/spoke/controllers/manifestcontroller/manifestwork_reconciler.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (zhiweiyin318) [reply pkg/work/spoke/controllers/manifestcontroller/manifestwork_reconciler.go]: changed to only pick up the first one for feedbackRule and installStrategy.

**Comment by zhiweiyin318** [comment]
@zhujian7 @qiujian16  please take a look again.

File: pkg/work/helper/helpers.go [hunk 4]
```diff
@@ -364,20 +365,18 @@ func NewAppliedManifestWorkOwner(appliedWork *workapiv1.AppliedManifestWork) *me
...
+func FindManifestConfiguration(resourceMeta workapiv1.ManifestResourceMeta,
+	manifestOptions []workapiv1.ManifestConfigOption) *workapiv1.ManifestConfigOption {
+	// return the first matched option
+	for i := 0; i < len(manifestOptions); i++ {
+		option := manifestOptions[i]
+		if resourceMeta.Group == option.ResourceIdentifier.Group &&
+			resourceMeta.Resource == option.ResourceIdentifier.Resource &&
+			wildcardMatch(resourceMeta.Namespace, option.ResourceIdentifier.Namespace) &&
+			wildcardMatch(resourceMeta.Name, option.ResourceIdentifier.Name) {
+			return &option
```
Comment (zhujian7) [inline pkg/work/helper/helpers.go:377]: 1. If two options match, the first configured the FeedbackRules but no UpdateStrategy, and the second configured the UpdateStrategy but no FeedbackRules, will returning the first option cause the FeedbackRules to be lost? 
2. I am still considering if we should make the word-by-word matching take precedence over wildcard match?

Comment (zhiweiyin318) [reply pkg/work/helper/helpers.go:399]: > If two options matched, the first configured the FeedbackRules but no UpdateStrategy, and the second configured the UpdateStrategy but no FeedbackRules, will returning the first option cause the FeedbackRules to be lost?

fixed by splitting the FindManifestConfiguration into 2 functions.

> I am still considering if we should make the word-by-word matching take precedence over wildcard matching?
I think the users only use the first option to filter resources in the most cases, so word-by-word matching is the same level as wildcard matching.
@qiujian16 WDYT?

Comment (qiujian16) [reply pkg/work/helper/helpers.go:399]: could we just build a new option finding the first available rule and the first updateStrategy in the FindManifestConfiguration?

Comment (zhiweiyin318) [reply pkg/work/helper/helpers.go]: done. changed to one FindManifestConfiguration and return an option with the first matched updateStrategy and feedbackRule.

**Review by qiujian16 — COMMENTED** [review]
/lgtm

**Pull request merged**